From 3007ca84bf61e7a9c831039c82e25e5e71f5c68f Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Wed, 31 Aug 2022 11:40:30 +0200
Subject: [PATCH 073/954] feat(stm32mp13): save/restore DDRMCE config

The DDRMCE regions configuration is saved in backup RAM context for
standby mode. This allows reloading the configuration without fetching
it from boot device.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 021532ce274e799c2160fb677e076457bfcf9659 feat(stm32mp13): save/restore DDRMCE config]
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Change-Id: I29e5f2f16d6a9d08bdaac1d7cc0fd3443a92e3b1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/264426
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
---
 plat/st/stm32mp1/include/stm32mp1_context.h |  6 +++-
 plat/st/stm32mp1/stm32mp1_context.c         | 34 +++++++++++++++++++++
 2 files changed, 39 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp1/include/stm32mp1_context.h b/plat/st/stm32mp1/include/stm32mp1_context.h
index 68bc3d3191..9b815023c9 100644
--- a/plat/st/stm32mp1/include/stm32mp1_context.h
+++ b/plat/st/stm32mp1/include/stm32mp1_context.h
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2021, ARM Limited and Contributors. All rights reserved.
+ * Copyright (c) 2017-2022, ARM Limited and Contributors. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -10,6 +10,8 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#include <drivers/st/stm32_mce.h>
+
 void stm32_clean_context(void);
 void stm32_context_save_bl2_param(void);
 uint32_t stm32_get_zdata_from_context(void);
@@ -19,6 +21,8 @@ uint32_t stm32_pm_get_optee_ep(void);
 #if STM32MP13
 void stm32mp1_pm_save_mce_mkey_in_context(uint8_t *data);
 void stm32mp1_pm_get_mce_mkey_from_context(uint8_t *data);
+void stm32mp1_pm_save_mce_region(uint32_t index, struct stm32_mce_region_s *config);
+void stm32mp1_pm_get_mce_region(uint32_t index, struct stm32_mce_region_s *config);
 #endif
 
 #endif /* STM32MP1_CONTEXT_H */
diff --git a/plat/st/stm32mp1/stm32mp1_context.c b/plat/st/stm32mp1/stm32mp1_context.c
index ce4bdc9475..c039fe0beb 100644
--- a/plat/st/stm32mp1/stm32mp1_context.c
+++ b/plat/st/stm32mp1/stm32mp1_context.c
@@ -63,6 +63,7 @@ struct backup_data_s {
 	uint32_t bl2_end;
 #if STM32MP13
 	uint8_t mce_mkey[MCE_KEY_SIZE_IN_BYTES];
+	struct stm32_mce_region_s mce_regions[MCE_IP_MAX_REGION_NB];
 #endif
 };
 
@@ -214,4 +215,37 @@ void stm32mp1_pm_get_mce_mkey_from_context(uint8_t *data)
 	clk_disable(BKPSRAM);
 }
 
+void stm32mp1_pm_save_mce_region(uint32_t index, struct stm32_mce_region_s *config)
+{
+	struct backup_data_s *backup_data;
+
+	if (index >= MCE_IP_MAX_REGION_NB) {
+		panic();
+	}
+
+	backup_data = (struct backup_data_s *)STM32MP_BACKUP_RAM_BASE;
+
+	clk_enable(BKPSRAM);
+
+	memcpy(&backup_data->mce_regions[index], config, sizeof(struct stm32_mce_region_s));
+
+	clk_disable(BKPSRAM);
+}
+
+void stm32mp1_pm_get_mce_region(uint32_t index, struct stm32_mce_region_s *config)
+{
+	struct backup_data_s *backup_data;
+
+	if (index >= MCE_IP_MAX_REGION_NB) {
+		panic();
+	}
+
+	backup_data = (struct backup_data_s *)STM32MP_BACKUP_RAM_BASE;
+
+	clk_enable(BKPSRAM);
+
+	memcpy(config, &backup_data->mce_regions[index], sizeof(struct stm32_mce_region_s));
+
+	clk_disable(BKPSRAM);
+}
 #endif
-- 
2.53.0

